pkg/controllers/jobset_controller_test.go

```diff
@@ -1205,3 +1217,152 @@ func makeJob(args *makeJobArgs) *testutils.JobWrapper {
+		jobSet                         *jobset.JobSet
+		existingService                bool
+		expectServiceCreate            bool
+		expectServiceName              string
+		expectPublishNotReadyAddresses bool
+		expectErr                      bool
+		expectErrStr                   string
+	}{
+		{
+			name:            "headless service exists and should not be created",
+			jobSet:          testutils.MakeJobSet(jobSetName, ns).EnableDNSHostnames(true).Obj(),
+			existingService: true,
+		},
+		{
+			name:            "headless service creation fails with unexpected error",
+			jobSet:          testutils.MakeJobSet(jobSetName, ns).EnableDNSHostnames(true).Obj(),
+			existingService: false,
+			expectErr:       true,
+			expectErrStr:    "unexpected error",
+		},
+		{
+			name:                           "service does not exist and should be created, subdomain not set",
+			jobSet:                         testutils.MakeJobSet(jobSetName, ns).EnableDNSHostnames(true).Obj(),
+			expectServiceCreate:            true,
+			expectServiceName:              "test-jobset",
+			expectPublishNotReadyAddresses: true,
+		},
+		{
+			name:                           "service does not exist and should be created, subdomain set",
+			jobSet:                         testutils.MakeJobSet(jobSetName, ns).EnableDNSHostnames(true).NetworkSubdomain("test-subdomain").Obj(),
+			expectServiceCreate:            true,
+			expectServiceName:              "test-subdomain",
+			expectPublishNotReadyAddresses: true,
+		},
+		{
+			name:                           "service does not exist and should be created, publishNotReadyAddresses is false",
+			jobSet:                         testutils.MakeJobSet(jobSetName, ns).EnableDNSHostnames(true).PublishNotReadyAddresses(false).Obj(),
+			expectServiceCreate:            true,
+			expectServiceName:              "test-jobset",
+			expectPublishNotReadyAddresses: false,
+		},
+		{
+			name:                           "service does not exist and should be created, publishNotReadyAddresses is true",
+			jobSet:                         testutils.MakeJobSet(jobSetName, ns).EnableDNSHostnames(true).PublishNotReadyAddresses(true).Obj(),
+			expectServiceCreate:            true,
+			expectServiceName:              "test-jobset",
+			expectPublishNotReadyAddresses: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var servicesCreated int
+			_, ctx := ktesting.NewTestContext(t)
+			scheme := runtime.NewScheme()
+			utilruntime.Must(jobset.AddToScheme(scheme))
+			utilruntime.Must(corev1.AddToScheme(scheme))
+			fakeClientBuilder := fake.NewClientBuilder().WithScheme(scheme).WithInterceptorFuncs(interceptor.Funcs{
+				Create: func(ctx context.Context, client client.WithWatch, obj client.Object, opts ...client.CreateOption) error {
+					if tc.expectErr {
+						return errors.New("unexpected error")
+					}
+					if !tc.expectServiceCreate {
+						t.Fatal("unexpected service creation")
```

Comment:
Why not t.Error() here? Don't we want to continue running the rest of the tests?

Reply:
I am not sure, it is an error if an unexpected service creation happens, but does it make sense to do further assertion on that service, if we didn't expect creaton in the first place?
At the end we also have assertions how many services we expected to be created (1 if we expect, 0 if we don't) which would also get skipped.

I don't have a strong opinion, you can make the decision.